src/Cli/dotnet/commands/VirtualProjectBuildingCommand.cs

```diff
@@ -0,0 +1,194 @@
+            {
+                return 1;
+            }
+
+            BuildManager.DefaultBuildManager.EndBuild();
+            return 0;
+        }
+        catch (Exception e)
+        {
+            Console.Error.WriteLine(e.Message);
+            return 1;
+        }
+        finally
+        {
+            foreach (var (key, value) in savedEnvironmentVariables)
+            {
+                Environment.SetEnvironmentVariable(key, value);
+            }
+
+            foreach (var binaryLogger in binaryLoggers)
+            {
+                binaryLogger.Shutdown();
+            }
+
+            consoleLogger.Shutdown();
+        }
+
+        static ImmutableArray<ILogger> GetBinaryLoggers(string[] args)
+        {
+            return args
+                .Where(RunCommand.IsBinLogArgument)
+                .Select(static ILogger (arg) => new BinaryLogger
+                {
+                    Parameters = arg.IndexOf(':') is >= 0 and var index
+                            ? arg[(index + 1)..]
+                            : "msbuild.binlog",
+                })
+                .ToImmutableArray();
+        }
+    }
+
+    public ProjectInstance CreateProjectInstance(ProjectCollection projectCollection)
+    {
+        return CreateProjectInstance(projectCollection, addGlobalProperties: null);
+    }
+
+    private ProjectInstance CreateProjectInstance(
+        ProjectCollection projectCollection,
+        Action<IDictionary<string, string>>? addGlobalProperties)
+    {
+        var projectRoot = CreateProjectRootElement(projectCollection);
+
+        var globalProperties = projectCollection.GlobalProperties;
+        if (addGlobalProperties is not null)
+        {
+            globalProperties = new Dictionary<string, string>(projectCollection.GlobalProperties, StringComparer.OrdinalIgnoreCase);
+            addGlobalProperties(globalProperties);
+        }
+
+        return ProjectInstance.FromProjectRootElement(projectRoot, new ProjectOptions
+        {
+            GlobalProperties = globalProperties,
+        });
+    }
+
+    private ProjectRootElement CreateProjectRootElement(ProjectCollection projectCollection)
+    {
+        var projectFileFullPath = Path.ChangeExtension(EntryPointFileFullPath, ".csproj");
+        var projectFileText = """
+            <Project>
+                <Import Project="Sdk.props" Sdk="Microsoft.NET.Sdk" />
+
+                <PropertyGroup>
+                    <OutputType>Exe</OutputType>
+                    <TargetFramework>net9.0</TargetFramework>
+                    <ImplicitUsings>enable</ImplicitUsings>
+                    <Nullable>enable</Nullable>
+                </PropertyGroup>
+
+                <Import Project="Sdk.targets" Sdk="Microsoft.NET.Sdk" />
+
+                <!--
+                    Override targets which don't work with project files that are not present on disk.
+                    See https://github.com/NuGet/Home/issues/14148.
```

Comment:
What if we call static graph restore?
Is this still a problem?
cc @jeffkl

Reply:
Static graph restore never triggers any of the target being reported as a problem, so I thought it'd just skip the problem.
Why do we have a concern with static graph?

Reply:
Just wanted to make sure you saw, why is static graph restore a concern?

Reply:
Oh sorry, I thought you were asking what if user enabled static graph restore, whether this will work. But you are saying we should enable static graph restore and then perhaps we could avoid this workaround. Thanks, I will investigate that.

Reply:
1. Looks like static graph restore fails similarly - cannot find the project file on disk.
2. Nevertheless, I'm not sure we would want to enable it by default, doesn't it have some differences? We would ideally want to have the "default" experience with `dotnet run file.cs`.
3. Even with the current workaround, if users explicitly enable static graph restore (e.g., using `-p:` argument or `Directory.Build.props`), the `dotnet run file.cs` will fail because some other NuGet targets cannot find the project file. I guess we could try to find workarounds for that as well, but I think we can just say that's not supported for now.

Reply:
1. I'd be curious to see the error that's happening and where it's getting thrown from. 
These targets are not invoked at all during static graph restore, so it's surprising if the error is the same. 
2. There's no functional differences between the 2. The biggest reason why static graph is not default is because it's slower for smaller projects because of the extra process.
3. What are we saying is not supported?

Reply:
> 1. I'd be curious to see the error that's happening and where it's getting thrown from.
>    These targets are not invoked at all during static graph restore, so it's surprising if the error is the same.

The error is not exactly the same, let me try to figure out where it's coming from.

> 3\. What are we saying is not supported?

Static graph restore in combination with `dotnet run file.cs`.

Reply:
I might be misunderstanding this scenario, but given that customers don't see a project file, I don't think they care about static graph vs regular restore, they just care that it gets compiled and run.

Reply:
Sure, but they could do `dotnet run file.cs -p:RestoreUseStaticGraphEvaluation=true` or more likely have `RestoreUseStaticGraphEvaluation` set in a parent `Directory.Build.props` and then if such `dotnet run file.cs` fails that seems like a bug (unless we say it's unsupported for starters).

Reply:
So the error comes from RestoreTaskEx, not sure where exactly but found this test that shows the exact error:

https://github.com/NuGet/NuGet.Client/blob/54f4fd0b4b418666683ab717d075fe0a239eb484/test/NuGet.Core.FuncTests/Msbuild.Integration.Test/MsbuildRestoreTaskTests.cs#L821

The error is `error MSB4025: The project file could not be loaded. Could not find file '...'`.

For repro, you can use this branch: https://github.com/jjonescz/MSBuildInMemory/tree/static-graph